### PR TITLE
Create hyperlinks component in GOVUK Publishing Components

### PR DESCRIPTION
**Pull request by DilwoarH**
[Rendered version](https://github.com/alphagov/govuk-rfcs/blob/add-components-link/rfc-127-create-links-component.md)

**Deadline:** 2020-06-30

### CODE REVIEW

**Comment by andysellick** [comment]
I'm wondering about the need for this - would the security vulnerability you mention be possible on GOV.UK? If we do have a security vulnerability relating to this problem we should prioritise fixing that first - wouldn't building a new component and then rolling it out to every link we have create more work for us unnecessarily?

I can't comment on having link styles on GOV.UK inconsistent with the Design System except to say we have a lot of things that aren't consistent with the DS, sometimes for good reason. Again, if there's instances where we should be and aren't, I think creating a new component first is just adding to the work.

There are also a lot of links that aren't just styled with the `govuk-link` class, both in applications and components ([example](https://github.com/alphagov/govuk_publishing_components/blob/master/app/views/govuk_publishing_components/components/_action_link.html.erb#L27), [another example](https://github.com/alphagov/govuk_publishing_components/blob/master/app/views/govuk_publishing_components/components/_attachment.html.erb#L55), and [another](https://github.com/alphagov/govuk_publishing_components/blob/master/app/views/govuk_publishing_components/components/_contents_list.html.erb#L32)). While some of those could be dealt with using options on a link component (inverse, for example) not all could. We recently reached an agreement that we wouldn't allow components to be passed styles as this breaks component isolation - how would this new component deal with that?

Thinking back I vaguely remember discussing whether we should make a link component when we first built the component gem and deciding not to. I can't remember the reasoning behind it but maybe @vanitabarrett or @maxgds may recall.

**Comment by chao-xian** [comment]
I agree with @vanitabarrett, it's my understanding that in general components in design systems are made up of elements, links being considered at such that granular level. I don't think this is the right approach either.

Also, [Govspeak won't allow you to set any attributes other than `data`](https://github.com/alphagov/govspeak/blob/master/lib/govspeak/html_sanitizer.rb#L57) (plus `role` and `aria-label`.

File: rfc-127-create-links-component.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+## Problem
+
+Links styles across GOVUK are not consistent and often security factors are not considered when using links within GOVUK. This is often due to the lack of frontend developers available in teams. 
+
+Sometimes security issues are not known by Developers and fixing them across multiple applications can add a lot of techincal debt - one recent example is the `target="_blank"` vulnerbility (https://www.manjuladube.dev/target-blank-security-vulnerability). Use of `target="_blank"` can be seen here: https://github.com/search?q=org%3Aalphagov+%27target%3D%22_blank%22%27&type=Code
+
+There are also certain styles that have been introduced in GOVUK that does not exist in the GOVUK Frontend Design system.
+
+<img src="https://raw.githubusercontent.com/alphagov/govuk-rfcs/add-components-link/rfc-127/Screenshot%202020-06-15%20at%2016.37.10.png" />
```
[the remark this answers is not in the document]
Comment (DilwoarH) [reply rfc-127-create-links-component.md:15]: Sorry - this is from Content publisher

[the remark this answers is not in the document]
Comment (DilwoarH) [reply rfc-127-create-links-component.md:15]: Screenshot 2020-06-16 at 10 25 50

**Comment by DilwoarH** [comment]
> I agree with @vanitabarrett, it's my understanding that in general components in design systems are made up of elements, links being considered at that granular level. Making a link a component doesn't fit into that model.
> 
> I'm not sure this is the appropriate approach to a solving a `target=_"blank"` vulnerability either.
> 
> Also, [Govspeak won't allow you to set any attributes other than `data`](https://github.com/alphagov/govspeak/blob/master/lib/govspeak/html_sanitizer.rb#L57) (plus `role` and `aria-label`.

Thank you for clarifying the comment @chao-xian and clarifying the GovSpeak issue (this was something that was of concern)

**Comment by DilwoarH** [comment]
Thank your @andysellick @bilbof @vanitabarrett @chao-xian - your feedback is really useful.

**Comment by DilwoarH** [comment]
> This sounds good to me. I was surprised we didn't already have one 👍

Hey @bilbof - could you expand a bit on why you agree? It will be good to understand from a backend developer prospective.

**Comment by bilbof** [comment]
@DilwoarH my agreement is from a position of ignorance about the the components.

The benefit to me would be that there would be a thing I could reach for when I need to create a link on a page, rather than needing to find the relevant classes for the element. I've searched the components guide before for a link component.

This wouldn't solve all problems, since occasionally - as mentioned above - it's tricky to use a component. Perhaps this could also be solved by documentation around the elements available.

**Comment by chao-xian** [comment]
@bilbof https://design-system.service.gov.uk/styles/typography/#links

**Comment by bilbof** [comment]
@chao-xian amazing! Thank you 🙏 

I don't usually check the design system site, but I will in future.

**Comment by kevindew** [comment]
Thanks for raising @DilwoarH 👍  Do you have an idea what the component code would look like? Just to consider how many things it does in order to justify whether it should exist?

With it being such a granular element it seems a tricky one in that it'd need to be super flexible to cover all link use cases, but if it's super flexible it might not benefit from being a component.

I wonder if there are other options too like helper functions that could be in between writing repetitive HTML vs having a rigid component?

> Also, [Govspeak won't allow you to set any attributes other than `data`](https://github.com/alphagov/govspeak/blob/master/lib/govspeak/html_sanitizer.rb#L57) (plus `role` and `aria-label`.

Argh, it really shouldn't let people set data 🙈

**Comment by DilwoarH** [comment]
> Thanks for raising @DilwoarH 👍 Do you have an idea what the component code would look like? Just to consider how many things it does in order to justify whether it should exist?
> 
> With it being such a granular element it seems a tricky one in that it'd need to be super flexible to cover all link use cases, but if it's super flexible it might not benefit from being a component.
> 
> I wonder if there are other options too like helper functions that could be in between writing repetitive HTML vs having a rigid component?
> 
> > Also, [Govspeak won't allow you to set any attributes other than `data`](https://github.com/alphagov/govspeak/blob/master/lib/govspeak/html_sanitizer.rb#L57) (plus `role` and `aria-label`.
> 
> Argh, it really shouldn't let people set data 🙈

Thanks for raising this - I will get some more information on this for you

**Comment by kevindew** [comment]
Since the deadline passed some time ago I'm going to close this for now. We can always re-open this again for further discussion should a fresh driver for this emerge..

**Pull request closed without merging**